warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ValidateFilterFunctionVisitor.java

```diff
@@ -0,0 +1,412 @@
+import static datawave.query.jexl.functions.EvaluationPhaseFilterFunctionsDescriptor.NO_EXPANSION;
+
+/**
+ * Verifies that no filter function is run against an index-only field
+ * <p>
+ * The exceptions to this rule are the #NO_EXPANSION and #TEXT functions.
+ * <p>
+ * The #NO_EXPANSION term will be pruned from the query prior to this visitor's execution, and the #TEXT query will have an index query added into the query.
+ */
+public class ValidateFilterFunctionVisitor extends BaseVisitor {
+    
+    private static final Logger log = ThreadConfigurableLogger.getLogger(ValidateFilterFunctionVisitor.class);
+    
+    private final Set<String> indexOnlyFields;
+    private final Set<String> excludedFunctions = Sets.newHashSet(NO_EXPANSION, INCLUDE_TEXT);
+    
+    public ValidateFilterFunctionVisitor(Set<String> indexOnlyFields) {
+        this.indexOnlyFields = indexOnlyFields;
+    }
+    
+    /**
+     * Generic entrypoint
+     *
+     * @param node
+     *            an arbitrary JexlNode
+     * @return the original node
+     */
+    public static JexlNode validate(JexlNode node, Set<String> indexOnlyFields) {
+        ValidateFilterFunctionVisitor visitor = new ValidateFilterFunctionVisitor(indexOnlyFields);
+        node.jjtAccept(visitor, null);
+        return node;
+    }
+    
+    @Override
+    public Object visit(ASTFunctionNode node, Object data) {
+        // visit children first, we may have function as an argument
+        node.childrenAccept(this, data);
+        validateFunction(node);
+        return data;
+    }
+    
+    private void validateFunction(ASTFunctionNode node) {
+        FunctionJexlNodeVisitor visitor = new FunctionJexlNodeVisitor();
+        node.jjtAccept(visitor, null);
+        
+        Set<String> identifiers;
+        if (visitor.namespace().equals(EVAL_PHASE_FUNCTION_NAMESPACE) && !excludedFunctions.contains(visitor.name())) {
+            for (JexlNode arg : visitor.args()) {
+                identifiers = JexlASTHelper.getIdentifierNames(arg);
+                for (String identifier : identifiers) {
+                    if (indexOnlyFields.contains(identifier)) {
+                        
+                        if (log.isDebugEnabled()) {
+                            log.debug("Filter function contained index-only field [" + identifier + "] for node ["
+                                            + JexlStringBuildingVisitor.buildQueryWithoutParse(node) + "]");
+                        }
+                        
+                        String msg = "Filter function cannot evaluate against index-only field [" + identifier + "]";
+                        BadRequestQueryException qe = new BadRequestQueryException(DatawaveErrorCode.INVALID_FUNCTION_ARGUMENTS, msg);
+                        throw new DatawaveFatalQueryException(qe);
+                    }
+                }
+            }
+        }
+    }
+    
+    // pass-through
+    
+    @Override
```

Comment:
I'm curious about why you're overriding these functions when pass-through is already the default behavior in BaseVisitor. Is it mainly to document which nodes are pass-through vs. not?

Reply:
That is the precisely the case -- these overrides exist to serve as documentation for which nodes the visitor will pass through/descend/visit.

Reply:
Okay, good to know.